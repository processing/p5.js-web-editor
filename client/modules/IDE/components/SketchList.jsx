import PropTypes from 'prop-types';
import React from 'react';
import { Helmet } from 'react-helmet';
import { withTranslation } from 'react-i18next';
import { connect } from 'react-redux';
import { Link } from 'react-router';
import { bindActionCreators } from 'redux';
import classNames from 'classnames';
import slugify from 'slugify';
import dates from '../../../utils/formatDate';
import * as ProjectActions from '../actions/project';
import * as ProjectsActions from '../actions/projects';
import * as CollectionsActions from '../actions/collections';
import * as ToastActions from '../actions/toast';
import * as SortingActions from '../actions/sorting';
import * as IdeActions from '../actions/ide';
import getSortedSketches from '../selectors/projects';
import Loader from '../../App/components/loader';
import Overlay from '../../App/components/Overlay';
import AddToCollectionList from './AddToCollectionList';

import ArrowUpIcon from '../../../images/sort-arrow-up.svg';
import ArrowDownIcon from '../../../images/sort-arrow-down.svg';
import DownFilledTriangleIcon from '../../../images/down-filled-triangle.svg';

const formatDateCell = (date, mobile = false) =>
  dates.format(date, { showTime: !mobile });

class SketchListRowBase extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      optionsOpen: false,
      renameOpen: false,
      renameValue: props.sketch.name,
      isFocused: false
    };
    this.renameInput = React.createRef();
  }

  onFocusComponent = () => {
    this.setState({ isFocused: true });
  };

  onBlurComponent = () => {
    this.setState({ isFocused: false });
    setTimeout(() => {
      if (!this.state.isFocused) {
        this.closeAll();
      }
    }, 200);
  };

  openOptions = () => {
    this.setState({
      optionsOpen: true
    });
  };

  closeOptions = () => {
    this.setState({
      optionsOpen: false
    });
  };

  toggleOptions = () => {
    if (this.state.optionsOpen) {
      this.closeOptions();
    } else {
      this.openOptions();
    }
  };

  openRename = () => {
    this.setState(
      {
        renameOpen: true,
        renameValue: this.props.sketch.name
      },
      () => this.renameInput.current.focus()
    );
  };

  closeRename = () => {
    this.setState({
      renameOpen: false
    });
  };

  closeAll = () => {
    this.setState({
      renameOpen: false,
      optionsOpen: false
    });
  };

  handleRenameChange = (e) => {
    this.setState({
      renameValue: e.target.value
    });
  };

  handleRenameEnter = (e) => {
    if (e.key === 'Enter') {
      this.updateName();
      this.closeAll();
    }
  };

  handleRenameBlur = () => {
    this.updateName();
    this.closeAll();
  };

  updateName = () => {
    const isValid = this.state.renameValue.trim().length !== 0;
    if (isValid) {
      this.props.changeProjectName(
        this.props.sketch.id,
        this.state.renameValue.trim()
      );
    }
  };

  resetSketchName = () => {
    this.setState({
      renameValue: this.props.sketch.name,
      renameOpen: false
    });
  };

  handleDropdownOpen = () => {
    this.closeAll();
    this.openOptions();
  };

  handleRenameOpen = () => {
    this.closeAll();
    this.openRename();
  };

  handleSketchDownload = () => {
    this.props.exportProjectAsZip(this.props.sketch.id);
  };

  handleSketchDuplicate = () => {
    this.closeAll();
    this.props.cloneProject(this.props.sketch);
  };

  handleSketchShare = () => {
    this.closeAll();
    this.props.showShareModal(
      this.props.sketch.id,
      this.props.sketch.name,
      this.props.username
    );
  };

  handleSketchDelete = () => {
    this.closeAll();
    if (
      window.confirm(
        this.props.t('Common.DeleteConfirmation', {
          name: this.props.sketch.name
        })
      )
    ) {
      this.props.deleteProject(this.props.sketch.id);
    }
  };

  renderViewButton = (sketchURL) => (
    <td className="sketch-list__dropdown-column">
      <Link to={sketchURL}>{this.props.t('SketchList.View')}</Link>
    </td>
  );

  renderDropdown = () => {
    const { optionsOpen } = this.state;
    const userIsOwner = this.props.user.username === this.props.username;

    return (
      <td className="sketch-list__dropdown-column">
        <button
          className="sketch-list__dropdown-button"
          onClick={this.toggleOptions}
          onBlur={this.onBlurComponent}
          onFocus={this.onFocusComponent}
          aria-label={this.props.t('SketchList.ToggleLabelARIA')}
        >
          <DownFilledTriangleIcon focusable="false" aria-hidden="true" />
        </button>
        {optionsOpen && (
          <ul className="sketch-list__action-dialogue">
            {userIsOwner && (
              <li>
                <button
                  className="sketch-list__action-option"
                  onClick={this.handleRenameOpen}
                  onBlur={this.onBlurComponent}
                  onFocus={this.onFocusComponent}
                >
                  {this.props.t('SketchList.DropdownRename')}
                </button>
              </li>
            )}
            <li>
              <button
                className="sketch-list__action-option"
                onClick={this.handleSketchDownload}
                onBlur={this.onBlurComponent}
                onFocus={this.onFocusComponent}
              >
                {this.props.t('SketchList.DropdownDownload')}
              </button>
            </li>
            {this.props.user.authenticated && (
              <li>
                <button
                  className="sketch-list__action-option"
                  onClick={this.handleSketchDuplicate}
                  onBlur={this.onBlurComponent}
                  onFocus={this.onFocusComponent}
                >
                  {this.props.t('SketchList.DropdownDuplicate')}
                </button>
              </li>
            )}
            {this.props.user.authenticated && (
              <li>
                <button
                  className="sketch-list__action-option"
                  onClick={() => {
                    this.props.onAddToCollection();
                    this.closeAll();
                  }}
                  onBlur={this.onBlurComponent}
                  onFocus={this.onFocusComponent}
                >
                  {this.props.t('SketchList.DropdownAddToCollection')}
                </button>
              </li>
            )}
            {/* <li>
              <button
                className="sketch-list__action-option"
                onClick={this.handleSketchShare}
                onBlur={this.onBlurComponent}
                onFocus={this.onFocusComponent}
              >
                Share
              </button>
            </li> */}
            {userIsOwner && (
              <li>
                <button
                  className="sketch-list__action-option"
                  onClick={this.handleSketchDelete}
                  onBlur={this.onBlurComponent}
                  onFocus={this.onFocusComponent}
                >
                  {this.props.t('SketchList.DropdownDelete')}
                </button>
              </li>
            )}
          </ul>
        )}
      </td>
    );
  };

  render() {
    const { sketch, username, mobile } = this.props;
    const { renameOpen, renameValue } = this.state;
    let url = `/${username}/sketches/${sketch.id}`;
    if (username === 'p5') {
      url = `/${username}/sketches/${slugify(sketch.name, '_')}`;
    }

    const name = (
      <React.Fragment>
        <Link to={url}>{renameOpen ? '' : sketch.name}</Link>
        {renameOpen && (
          <input
            value={renameValue}
            onChange={this.handleRenameChange}
            onKeyUp={this.handleRenameEnter}
            onBlur={this.handleRenameBlur}
            onClick={(e) => e.stopPropagation()}
            ref={this.renameInput}
          />
        )}
      </React.Fragment>
    );

    return (
      <React.Fragment>
        <tr
          className="sketches-table__row"
          key={sketch.id}
          onClick={this.handleRowClick}
        >
          <th scope="row">{name}</th>
          <td>
            {mobile && 'Created: '}
            {formatDateCell(sketch.createdAt, mobile)}
          </td>
          <td>
            {mobile && 'Updated: '}
            {formatDateCell(sketch.updatedAt, mobile)}
          </td>
          {this.renderDropdown()}
        </tr>
      </React.Fragment>
    );
  }
}

SketchListRowBase.propTypes = {
  sketch: PropTypes.shape({
    id: PropTypes.string.isRequired,
    name: PropTypes.string.isRequired,
    createdAt: PropTypes.string.isRequired,
    updatedAt: PropTypes.string.isRequired
  }).isRequired,
  username: PropTypes.string.isRequired,
  user: PropTypes.shape({
    username: PropTypes.string,
    authenticated: PropTypes.bool.isRequired
  }).isRequired,
  deleteProject: PropTypes.func.isRequired,
  showShareModal: PropTypes.func.isRequired,
  cloneProject: PropTypes.func.isRequired,
  exportProjectAsZip: PropTypes.func.isRequired,
  changeProjectName: PropTypes.func.isRequired,
  onAddToCollection: PropTypes.func.isRequired,
  mobile: PropTypes.bool,
  t: PropTypes.func.isRequired
};

SketchListRowBase.defaultProps = {
  mobile: false
};

function mapDispatchToPropsSketchListRow(dispatch) {
  return bindActionCreators(
    Object.assign({}, ProjectActions, IdeActions),
    dispatch
  );
}

const SketchListRow = connect(
  null,
  mapDispatchToPropsSketchListRow
)(SketchListRowBase);

class SketchList extends React.Component {
  constructor(props) {
    super(props);
    this.props.getProjects(this.props.username);
    this.props.resetSorting();

    this.state = {
      isInitialDataLoad: true
    };
  }

  componentDidUpdate(prevProps) {
    if (
      this.props.sketches !== prevProps.sketches &&
      Array.isArray(this.props.sketches)
    ) {
      // eslint-disable-next-line react/no-did-update-set-state
      this.setState({
        isInitialDataLoad: false
      });
    }
  }

  getSketchesTitle() {
    if (this.props.username === this.props.user.username) {
      return this.props.t('SketchList.Title');
    }
    return this.props.t('SketchList.AnothersTitle', {
      anotheruser: this.props.username
    });
  }

  hasSketches() {
    return !this.isLoading() && this.props.sketches.length > 0;
  }

  isLoading() {
    return this.props.loading && this.state.isInitialDataLoad;
  }

  _renderLoader() {
    if (this.isLoading()) return <Loader />;
    return null;
  }

  _renderEmptyTable() {
    if (!this.isLoading() && this.props.sketches.length === 0) {
      return (
        <p className="sketches-table__empty">
          {this.props.t('SketchList.NoSketches')}
        </p>
      );
    }
    return null;
  }

  _getButtonLabel = (fieldName, displayName) => {
    const { field, direction } = this.props.sorting;
    let buttonLabel;
    if (field !== fieldName) {
      if (field === 'name') {
        buttonLabel = this.props.t('SketchList.ButtonLabelAscendingARIA', {
          displayName
        });
      } else {
        buttonLabel = this.props.t('SketchList.ButtonLabelDescendingARIA', {
          displayName
        });
      }
    } else if (direction === SortingActions.DIRECTION.ASC) {
      buttonLabel = this.props.t('SketchList.ButtonLabelDescendingARIA', {
        displayName
      });
    } else {
      buttonLabel = this.props.t('SketchList.ButtonLabelAscendingARIA', {
        displayName
      });
    }
    return buttonLabel;
  };

  _renderFieldHeader = (fieldName, displayName) => {
    const { field, direction } = this.props.sorting;
    const headerClass = classNames({
      'sketches-table__header': true,
      'sketches-table__header--selected': field === fieldName
    });
    const buttonLabel = this._getButtonLabel(fieldName, displayName);
    return (
      <th scope="col">
        <button
          className="sketch-list__sort-button"
          onClick={() => this.props.toggleDirectionForField(fieldName)}
          aria-label={buttonLabel}
        >
          <span className={headerClass}>{displayName}</span>
          {field === fieldName &&
            direction === SortingActions.DIRECTION.ASC && (
              <ArrowUpIcon
                role="img"
                aria-label={this.props.t('SketchList.DirectionAscendingARIA')}
                focusable="false"
              />
            )}
          {field === fieldName &&
            direction === SortingActions.DIRECTION.DESC && (
              <ArrowDownIcon
                role="img"
                aria-label={this.props.t('SketchList.DirectionDescendingARIA')}
                focusable="false"
              />
            )}
        </button>
      </th>
    );
  };

  render() {
    const username =
      this.props.username !== undefined
        ? this.props.username
        : this.props.user.username;
    const { mobile } = this.props;
    return (
      <article className="sketches-table-container">
        <Helmet>
          <title>{this.getSketchesTitle()}</title>
        </Helmet>
        {this._renderLoader()}
        {this._renderEmptyTable()}
        {this.hasSketches() && (
          <table
            className="sketches-table"
            summary={this.props.t('SketchList.TableSummary')}
          >
            <thead>
              <tr>
                {this._renderFieldHeader(
                  'name',
                  this.props.t('SketchList.HeaderName')
                )}
                {this._renderFieldHeader(
                  'createdAt',
                  this.props.t('SketchList.HeaderCreatedAt', {
                    context: mobile ? 'mobile' : ''
                  })
                )}
                {this._renderFieldHeader(
                  'updatedAt',
                  this.props.t('SketchList.HeaderUpdatedAt', {
                    context: mobile ? 'mobile' : ''
                  })
                )}
                <th scope="col"></th>
              </tr>
            </thead>
            <tbody>
              {this.props.sketches.map((sketch) => (
                <SketchListRow
                  mobile={mobile}
                  key={sketch.id}
                  sketch={sketch}
                  user={this.props.user}
                  username={username}
                  onAddToCollection={() => {
                    this.setState({ sketchToAddToCollection: sketch });
                  }}
                  t={this.props.t}
                />
              ))}
            </tbody>
          </table>
        )}
        {this.state.sketchToAddToCollection && (
          <Overlay
            isFixedHeight
            title={this.props.t('SketchList.AddToCollectionOverlayTitle')}
            closeOverlay={() =>
              this.setState({ sketchToAddToCollection: null })
            }
          >
            <AddToCollectionList
              project={this.state.sketchToAddToCollection}
              username={this.props.username}
              user={this.props.user}
            />
          </Overlay>
        )}
      </article>
    );
  }
}

SketchList.propTypes = {
  user: PropTypes.shape({
    username: PropTypes.string,
    authenticated: PropTypes.bool.isRequired
  }).isRequired,
  getProjects: PropTypes.func.isRequired,
  sketches: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      createdAt: PropTypes.string.isRequired,
      updatedAt: PropTypes.string.isRequired
    })
  ).isRequired,
  username: PropTypes.string,
  loading: PropTypes.bool.isRequired,
  toggleDirectionForField: PropTypes.func.isRequired,
  resetSorting: PropTypes.func.isRequired,
  sorting: PropTypes.shape({
    field: PropTypes.string.isRequired,
    direction: PropTypes.string.isRequired
  }).isRequired,
  mobile: PropTypes.bool,
  t: PropTypes.func.isRequired
};

SketchList.defaultProps = {
  username: undefined,
  mobile: false
};

function mapStateToProps(state) {
  return {
    user: state.user,
    sketches: getSortedSketches(state),
    sorting: state.sorting,
    loading: state.loading,
    project: state.project
  };
}

function mapDispatchToProps(dispatch) {
  return bindActionCreators(
    Object.assign(
      {},
      ProjectsActions,
      CollectionsActions,
      ToastActions,
      SortingActions
    ),
    dispatch
  );
}

export default withTranslation()(
  connect(mapStateToProps, mapDispatchToProps)(SketchList)
);
