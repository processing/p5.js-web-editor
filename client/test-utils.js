/**
 * This file re-exports @testing-library but ensures that
 * any calls to render have translations and theme available.
 *
 * This means tested components will be able to call
 * `t()` and have the translations of the default
 * language also components will be able to call
 * `prop()` and have the theming of the default theme.
 *
 * For i18n see: https://react.i18next.com/misc/testing#testing-without-stubbing
 */

// eslint-disable-next-line import/no-extraneous-dependencies
import { render } from '@testing-library/react';
import React from 'react';
import PropTypes from 'prop-types';
import { createStore } from 'redux';
import { Provider } from 'react-redux';

import { I18nextProvider } from 'react-i18next';
import { ThemeProvider as StyledThemeProvider } from 'styled-components';

import i18n from './i18n-test';
import rootReducer from './reducers';
import ThemeProvider from './modules/App/components/ThemeProvider';
import theme, { Theme } from './theme';

// re-export everything
// eslint-disable-next-line import/no-extraneous-dependencies
export * from '@testing-library/react';

const Providers = ({ children }) => (
  // eslint-disable-next-line react/jsx-filename-extension
  <StyledThemeProvider theme={{ ...theme[Theme.light] }}>
    <I18nextProvider i18n={i18n}>{children}</I18nextProvider>
  </StyledThemeProvider>
);

Providers.propTypes = {
  children: PropTypes.element.isRequired
};

function reduxRender(
  ui,
  {
    initialState,
    store = createStore(rootReducer, initialState),
    ...renderOptions
  } = {}
) {
  function Wrapper({ children }) {
    return (
      <I18nextProvider i18n={i18n}>
        <Provider store={store}>
          <ThemeProvider>{children}</ThemeProvider>
        </Provider>
      </I18nextProvider>
    );
  }

  Wrapper.propTypes = {
    children: PropTypes.element.isRequired
  };

  return render(ui, { wrapper: Wrapper, ...renderOptions });
}

const customRender = (ui, options) =>
  render(ui, { wrapper: Providers, ...options });

// override render method
export { customRender as render, reduxRender };
