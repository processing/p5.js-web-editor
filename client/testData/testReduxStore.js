const mockProjects = [
  {
    name: 'testsketch1',
    _id: 'testid1',
    updatedAt: '2021-02-26T04:58:29',
    files: [],
    createdAt: '2021-02-26T04:58:14',
    id: 'testid1'
  },
  {
    name: 'testsketch2',
    _id: 'testid2',
    updatedAt: '2021-02-23T17:40:43',
    files: [],
    createdAt: '2021-02-23T17:40:43',
    id: 'testid2'
  }
];

const initialTestState = {
  ide: {
    isPlaying: false,
    isAccessibleOutputPlaying: false,
    modalIsVisible: false,
    sidebarIsExpanded: false,
    consoleIsExpanded: true,
    preferencesIsVisible: false,
    projectOptionsVisible: false,
    newFolderModalVisible: false,
    uploadFileModalVisible: false,
    shareModalVisible: false,
    shareModalProjectId: 'abcd',
    shareModalProjectName: 'My Cute Sketch',
    shareModalProjectUsername: 'p5_user',
    keyboardShortcutVisible: false,
    unsavedChanges: false,
    infiniteLoop: false,
    previewIsRefreshing: false,
    infiniteLoopMessage: '',
    justOpenedProject: false,
    previousPath: '/',
    errorType: undefined,
    runtimeErrorWarningVisible: true,
    parentId: undefined
  },
  files: [
    {
      name: 'root',
      id: '606fc1c46045e19ca2ee2648',
      _id: '606fc1c46045e19ca2ee2648',
      children: [
        '606fc1c46045e19ca2ee2646',
        '606fc1c46045e19ca2ee2645',
        '606fc1c46045e19ca2ee2647'
      ],
      fileType: 'folder',
      content: ''
    },
    {
      name: 'sketch.js',
      content:
        'function setup() { createCanvas(400, 400); } function draw() { background(220); }',
      id: '606fc1c46045e19ca2ee2645',
      _id: '606fc1c46045e19ca2ee2645',
      isSelectedFile: true,
      fileType: 'file',
      children: []
    },
    {
      name: 'index.html',
      content: `<!DOCTYPE html> <html lang="en"> <head> <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.4.0/p5.js"></script> <script src="https://cdnjs.cloudflare.com/ajax/libs/p5.js/1.4.0/addons/p5.sound.min.js"></script> <link rel="stylesheet" type="text/css" href="style.css"> <meta charset="utf-8" /> </head> <body> <main> </main> <script src="sketch.js"></script> </body> </html>`,
      id: '606fc1c46045e19ca2ee2646',
      _id: '606fc1c46045e19ca2ee2646',
      fileType: 'file',
      children: []
    },
    {
      name: 'style.css',
      content:
        'html, body { margin: 0; padding: 0; } canvas { display: block; } ',
      id: '606fc1c46045e19ca2ee2647',
      _id: '606fc1c46045e19ca2ee2647',
      fileType: 'file',
      children: []
    }
  ],
  preferences: {
    fontSize: 18,
    autosave: true,
    linewrap: true,
    lineNumbers: true,
    lintWarning: false,
    textOutput: false,
    gridOutput: false,
    theme: 'light',
    autorefresh: false,
    language: 'en-US',
    autocloseBracketsQuotes: true
  },
  user: {
    email: 'happydog@example.com',
    username: 'happydog',
    preferences: {},
    apiKeys: [],
    verified: 'sent',
    id: '123456789',
    totalSize: 0,
    authenticated: true
  },
  project: {
    name: 'Zealous sunflower',
    updatedAt: '',
    isSaving: false
  },
  sketches: mockProjects,
  search: {
    collectionSearchTerm: '',
    sketchSearchTerm: ''
  },
  sorting: {
    field: 'createdAt',
    direction: 'DESCENDING'
  },
  editorAccessibility: {
    lintMessages: [],
    forceDesktop: false
  },
  toast: {
    isVisible: false,
    text: ''
  },
  console: [],
  assets: {
    list: [],
    totalSize: 0
  },
  loading: false,
  collections: []
};

export { mockProjects, initialTestState };
