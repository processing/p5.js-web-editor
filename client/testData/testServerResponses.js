// test data to use with msw

export const sketchResponse = {};

export const userResponse = {
  email: 'happydog@example.com',
  username: 'happydog',
  preferences: {
    fontSize: 18,
    lineNumbers: true,
    indentationAmount: 2,
    isTabIndent: false,
    autosave: true,
    linewrap: true,
    lintWarning: false,
    textOutput: false,
    gridOutput: false,
    theme: 'light',
    autorefresh: false,
    language: 'en-US',
    autocloseBracketsQuotes: true
  },
  apiKeys: [],
  verified: 'verified',
  id: 'testid',
  totalSize: 0,
  github: 'githubusername',
  google: 'googleusername'
};
