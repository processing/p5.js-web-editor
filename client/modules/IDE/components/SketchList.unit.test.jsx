import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { setupServer } from 'msw/node';
import { rest } from 'msw';
import { act } from 'react-dom/test-utils';
import SketchList from './SketchList';
import { reduxRender, fireEvent, screen, within } from '../../../test-utils';
import { initialTestState } from '../../../testData/testReduxStore';

jest.mock('../../../i18n');

const server = setupServer(
  rest.get(`/${initialTestState.user.username}/projects`, (req, res, ctx) =>
    // it just needs to return something so it doesn't throw an error
    // Sketchlist tries to grab projects on creation but for the unit test
    // we just feed those in as part of the initial state
    res(ctx.json({ data: 'foo' }))
  )
);

beforeAll(() => server.listen());
afterEach(() => server.resetHandlers());
afterAll(() => server.close());

describe('<Sketchlist />', () => {
  const mockStore = configureStore([thunk]);
  const store = mockStore(initialTestState);

  let subjectProps = { username: initialTestState.user.username };

  const subject = () =>
    reduxRender(<SketchList {...subjectProps} />, { store });

  afterEach(() => {
    store.clearActions();
  });

  it('has sample projects', () => {
    act(() => {
      subject();
    });
    expect(screen.getByText('testsketch1')).toBeInTheDocument();
    expect(screen.getByText('testsketch2')).toBeInTheDocument();
  });

  it('clicking on date created row header dispatches a reordering action', () => {
    act(() => {
      subject();
    });

    act(() => {
      fireEvent.click(screen.getByText(/date created/i));
    });

    const expectedAction = [{ type: 'TOGGLE_DIRECTION', field: 'createdAt' }];
    expect(store.getActions()).toEqual(expect.arrayContaining(expectedAction));
  });

  it('clicking on dropdown arrow opens sketch options - sketches belong to user', () => {
    act(() => {
      subject();
    });

    const row = screen.getByRole('row', {
      name: /testsketch1/
    });

    const dropdown = within(row).getByRole('button', {
      name: 'Toggle Open/Close Sketch Options'
    });

    act(() => {
      fireEvent.click(dropdown);
    });

    expect(screen.queryByText('Rename')).toBeInTheDocument();
    expect(screen.queryByText('Duplicate')).toBeInTheDocument();
    expect(screen.queryByText('Download')).toBeInTheDocument();
    expect(screen.queryByText('Add to collection')).toBeInTheDocument();
    expect(screen.queryByText('Delete')).toBeInTheDocument();
  });

  it('snapshot testing', () => {
    const { asFragment } = subject();
    expect(asFragment()).toMatchSnapshot();
  });

  describe('different user than the one who created the sketches', () => {
    beforeAll(() => {
      subjectProps = { username: 'notthesameusername' };
    });

    it('clicking on dropdown arrow opens sketch options without Rename or Delete option', () => {
      act(() => {
        subject();
      });

      const row = screen.getByRole('row', {
        name: /testsketch1/
      });

      const dropdown = within(row).getByRole('button', {
        name: 'Toggle Open/Close Sketch Options'
      });

      act(() => {
        fireEvent.click(dropdown);
      });

      expect(screen.queryByText('Rename')).not.toBeInTheDocument();
      expect(screen.queryByText('Duplicate')).toBeInTheDocument();
      expect(screen.queryByText('Download')).toBeInTheDocument();
      expect(screen.queryByText('Add to collection')).toBeInTheDocument();
      expect(screen.queryByText('Delete')).not.toBeInTheDocument();
    });
  });
});
