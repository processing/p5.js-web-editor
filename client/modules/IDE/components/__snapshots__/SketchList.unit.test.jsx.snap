// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sketchlist /> snapshot testing 1`] = `
<DocumentFragment>
  <article
    class="sketches-table-container"
  >
    <table
      class="sketches-table"
      summary="table containing all saved projects"
    >
      <thead>
        <tr>
          <th
            scope="col"
          >
            <button
              aria-label="Sort by Sketch descending."
              class="sketch-list__sort-button"
            >
              <span
                class="sketches-table__header"
              >
                Sketch
              </span>
            </button>
          </th>
          <th
            scope="col"
          >
            <button
              aria-label="Sort by Date Created ascending."
              class="sketch-list__sort-button"
            >
              <span
                class="sketches-table__header sketches-table__header--selected"
              >
                Date Created
              </span>
              <test-file-stub
                aria-label="Descending"
                focusable="false"
                role="img"
              />
            </button>
          </th>
          <th
            scope="col"
          >
            <button
              aria-label="Sort by Date Updated descending."
              class="sketch-list__sort-button"
            >
              <span
                class="sketches-table__header"
              >
                Date Updated
              </span>
            </button>
          </th>
          <th
            scope="col"
          />
        </tr>
      </thead>
      <tbody>
        <tr
          class="sketches-table__row"
        >
          <th
            scope="row"
          >
            <a>
              testsketch1
            </a>
          </th>
          <td>
            Feb 26, 2021, 4:58:14 AM
          </td>
          <td>
            Feb 26, 2021, 4:58:29 AM
          </td>
          <td
            class="sketch-list__dropdown-column"
          >
            <button
              aria-label="Toggle Open/Close Sketch Options"
              class="sketch-list__dropdown-button"
            >
              <test-file-stub
                aria-hidden="true"
                focusable="false"
              />
            </button>
          </td>
        </tr>
        <tr
          class="sketches-table__row"
        >
          <th
            scope="row"
          >
            <a>
              testsketch2
            </a>
          </th>
          <td>
            Feb 23, 2021, 5:40:43 PM
          </td>
          <td>
            Feb 23, 2021, 5:40:43 PM
          </td>
          <td
            class="sketch-list__dropdown-column"
          >
            <button
              aria-label="Toggle Open/Close Sketch Options"
              class="sketch-list__dropdown-button"
            >
              <test-file-stub
                aria-hidden="true"
                focusable="false"
              />
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </article>
</DocumentFragment>
`;
